API/projectClass/backgroundsClass.m

```diff
@@ -152,16 +171,71 @@
             
             if ~isempty(inputBlock.type)
                 inputBlock.type = validateOption(inputBlock.type, 'allowedTypes', obj.invalidTypeMessage).value;
+
+                % If the type of the background is changed, clear all
+                % source and value fields
+                if ~strcmpi(inputBlock.type, obj.backgrounds.varTable{row, 2})
+                    warning("warnings:fieldsCleared", "When changing the type of a background all unset fields are cleared");
+                    for i = 3:width(obj.backgrounds.varTable)
+                        obj.backgrounds.setValue(row, i, '');
+                    end
+
+                    % Having cleared the table, we need to re-parse the
+                    % inputs to the function
+                    p = inputParser;
+                    addParameter(p, 'name', obj.backgrounds.varTable{row, 1}, @isText);
+                    addParameter(p, 'type', obj.backgrounds.varTable{row, 2}, @(x) isText(x) || isenum(x));
+                    addParameter(p, 'source', obj.backgrounds.varTable{row, 3}, @isText);
+                    addParameter(p, 'value1', obj.backgrounds.varTable{row, 4}, @isText);
+                    addParameter(p, 'value2', obj.backgrounds.varTable{row, 5}, @isText);
+                    addParameter(p, 'value3', obj.backgrounds.varTable{row, 6}, @isText);
+                    addParameter(p, 'value4', obj.backgrounds.varTable{row, 7}, @isText);
+                    addParameter(p, 'value5', obj.backgrounds.varTable{row, 8}, @isText);
+
+                    parse(p, varargin{:});
+                    inputBlock = p.Results;
+                end
+
                 obj.backgrounds.setValue(row, 2, inputBlock.type);
             end
-            values = {inputBlock.value1, inputBlock.value2, inputBlock.value3, inputBlock.value4};
-            for i = 1:4
+
+            % For data and function types, source is the data/function name
+            source = convertStringsToChars(inputBlock.source);
+            if ~isempty(source)
+                if strcmpi(inputBlock.type, allowedTypes.Constant.value)
+                    source = obj.validateParam(source, obj.backgroundParams.getNames(), 'Background Param');
+                    maxValues = 0;
```

Comment:
`maxValues` is used in multiple places might be worth making a private/constant struct with these values are reusing that